xicam/catalog_viewer/__init__.py

```diff
@@ -50,43 +51,58 @@ def field_changed(self, new_device):
-            image_fields = []
-            for field, field_descriptor in all_fields.items():
-                field_shape = len(field_descriptor['shape'])
-                if field_shape > 1 and field_shape < 5:
-                    # if field contains at least 1 entry that is at least one-dimensional (shape=2) 
-                    # or 2-dimensional (shape=3) or up to 3-dimensional (shape=4)
-                    # then add field e.g. 'fccd_image'
-                    image_fields.append(field)
-
-            stream_fields = []
-            for stream in list(run_catalog):
-                # check if image_field exists in different streams
-                # if yes -> add to stream_fields -> can be selected in CatalogViewer select_stream
-                for image in image_fields:
-                    if image in run_catalog[stream].metadata['descriptors'][0]['data_keys'] and stream not in stream_fields:
-                        stream_fields.append(stream)
-                    else:
-                        msg.showMessage("{0} stream does not contain {1} image data".format(stream, image_fields[0]))
-                #ToDo:
-                # Where/how to display non 2D stream fields --> such as metadata?
-
+            # try and startup with primary catalog and whatever fields it has
+            filtered_stream_fields = get_fields_for_stream(self.stream_fields, 'primary')
+            self.stream_combo_box.clear()
+            self.stream_combo_box.addItems(get_all_streams(run_catalog))
             self.field_combo_box.clear()
-            self.field_combo_box.addItems(image_fields)
-            self.field_combo_box_stream.clear()
-            self.field_combo_box_stream.addItems(stream_fields)
+            self.field_combo_box.addItems(filtered_stream_fields)
+
         except Exception as e:
             msg.logError(e)
             msg.showMessage("Unable to display: ", str(e))
 
     def appendHeader(self):
         ...
+
+
+def get_stream_data_keys(run_catalog, stream):
+    return run_catalog[stream].metadata['descriptors'][0]['data_keys']
+
+
+def get_fields_for_stream(stream_fields, stream):
+    return [stream_field_tuple[1] for stream_field_tuple in stream_fields if stream_field_tuple[0] == stream]
+
+
+def get_all_streams(run_catalog):
+    return list(run_catalog)
+
+
+def get_all_image_fields(run_catalog):
+    all_fields = []
+    for stream in list(run_catalog):
+        stream_fields = get_stream_data_keys(run_catalog, stream)
+        field_names = stream_fields.keys()
+        for field_name in field_names:
+            field_shape = len(stream_fields[field_name]['shape'])
+            if field_shape > 1 and field_shape < 5:
+                # if field contains at least 1 entry that is at least one-dimensional (shape=2) 
+                # or 2-dimensional (shape=3) or up to 3-dimensional (shape=4)
+                # then add field e.g. 'fccd_image'
+                stream_field = (stream, field_name)
```

Comment:
Why did you decide to use a tuple here? Can we access the stream entry from each tuple in the list in line 71? Could we also use a dict for this here?